feat: add errors to wrong mssql configuration

saiprabhu-dandanayak opened the pull request:
## Description


Fixes https://github.com/appsmithorg/appsmith/issues/24726

## Automation

/ok-to-test tags=""

### :mag: Cypress test results
<!-- This is an auto-generated comment: Cypress test results  -->
> [!CAUTION]  
> If you modify the content in this section, you are likely to disrupt the CI result for your PR.

<!-- end of auto-generated comment: Cypress test results  -->


## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced validation for SQL Server datasource configurations, including connection mode, endpoints, and authentication details.

- **Bug Fixes**
  - Improved error messages for missing hostname and database name in SQL Server datasource configurations.

- **Tests**
  - Added tests to ensure proper validation of datasources with missing or null credentials, endpoints, hosts, and authentication details.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

saiprabhu-dandanayak commented:
> This PR has not seen activitiy for a while. It will be closed in 7 days unless further activity is detected.

Hi @rohan-arthur 
I hope you're well. I’m curious and excited to hear your thoughts on my pull request raised a while ago. Looking forward to your review!
Thank you.

akshayvijayjain commented:
Hi @Nikhil-Nandagopal ,

this pr is also pending from long if it is possible to prioritize

rohan-arthur commented:
@saiprabhu-dandanayak I have created a shadow PR for running some automated tests and checks. 
PR link: https://github.com/appsmithorg/appsmith/pull/34877 

1. spotless checks are failing, [see this link](https://github.com/appsmithorg/appsmith/actions/runs/9890786991/job/27319805377?pr=34877). 

2. server unit tests are running currently, you can check the status [here](https://github.com/appsmithorg/appsmith/actions/runs/9890786991/job/27319805610?pr=34877).

3. integration tests are running [here](https://github.com/appsmithorg/appsmith/actions/runs/9890861555).

saiprabhu-dandanayak commented:
> @saiprabhu-dandanayak I have created a shadow PR for running some automated tests and checks. PR link: #34877
> 
> 1. spotless checks are failing, [see this link](https://github.com/appsmithorg/appsmith/actions/runs/9890786991/job/27319805377?pr=34877).
> 2. server unit tests are running currently, you can check the status [here](https://github.com/appsmithorg/appsmith/actions/runs/9890786991/job/27319805610?pr=34877).
> 3. integration tests are running [here](https://github.com/appsmithorg/appsmith/actions/runs/9890861555).

Hi @rohan-arthur , fixed testcases by running `mvn spotless:apply` as suggested in the error logs , and verified through running testcases.
### Testcase screenshot
![image](https://github.com/appsmithorg/appsmith/assets/122865888/0fe2a6ee-0bbe-455c-ab5d-80d85a30e355)

rishabhrathod01 commented:
@saiprabhu-dandanayak there is a server unit test failing for the above changes. Could you check this  https://github.com/appsmithorg/appsmith/pull/34877#issuecomment-2231599631

saiprabhu-dandanayak commented:
Hi @rishabhrathod01 , I have checked it twice , and u have attached a screenshot for the same , there was an error in mssqlpugin testcases due to missing imports statements , but I have fixed it , will recheck it again.

rishabhrathod01 commented:
@saiprabhu-dandanayak won't be required. That was a false alert by the initial CI run. All the tests have passed now.  https://github.com/appsmithorg/appsmith/pull/34877

saiprabhu-dandanayak commented:
> @saiprabhu-dandanayak won't be required. That was a false alert by the initial CI run. All the tests have passed now. #34877

> Hi @rishabhrathod01 , took recent changes from release branch and updated this branch.

sneha122 reviewed (commented):
@saiprabhu-dandanayak Why is it showing changes in these two files?
JSPlugin$JSPluginExecutor.class and JSPlugin.class? I don't think these changes should be part of the PR

saiprabhu-dandanayak commented:
> @saiprabhu-dandanayak Why is it showing changes in these two files? JSPlugin$JSPluginExecutor.class and JSPlugin.class? I don't think these changes should be part of the PR

Hi @sneha122 , deleted those files

sneha122 commented on app/server/appsmith-plugins/mssqlPlugin/src/main/java/com/external/plugins/MssqlPlugin.java:
```diff
@@ -363,30 +364,47 @@ public void datasourceDestroy(HikariDataSource connection) {
+                invalids.add(MssqlErrorMessages.DS_MISSING_ENDPOINT_ERROR_MSG);
+            } else if (!isEmpty(datasourceConfiguration.getEndpoints())) {
+                for (final Endpoint endpoint : datasourceConfiguration.getEndpoints()) {
+                    if (endpoint.getHost() == null || endpoint.getHost().isBlank()) {
```
Can we replace `endpoint.getHost() == null || endpoint.getHost().isBlank()` with StringUtils.isBlank(endpoint.getHost()) as this will check for "", null or whitespaces (basically all cases)?

sneha122 replied:
https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#isBlank-java.lang.CharSequence-

akshayvijayjain replied:
As far as I remember, this was purposely done by @Shirisha-Manthoju ,

tagged, her FYI

saiprabhu-dandanayak replied:
Resolved

sneha122 commented on app/server/appsmith-plugins/mssqlPlugin/src/main/java/com/external/plugins/MssqlPlugin.java:
```diff
@@ -363,30 +364,47 @@ public void datasourceDestroy(HikariDataSource connection) {
-
             } else {
-                if (StringUtils.isEmpty(auth.getUsername())) {
+                if (auth.getUsername() == null || auth.getUsername().isBlank()) {
```
Can we replace these two checks with StringUtils.isBlank instead?

saiprabhu-dandanayak replied:
Resolved

sneha122 commented on app/server/appsmith-plugins/mssqlPlugin/src/main/java/com/external/plugins/MssqlPlugin.java:
```diff
@@ -363,30 +364,47 @@ public void datasourceDestroy(HikariDataSource connection) {
                 }
-
-                if (StringUtils.isEmpty(auth.getPassword())) {
+                if (auth.getPassword() == null || auth.getPassword().isBlank()) {
```
Same as above

saiprabhu-dandanayak replied:
Resolved

sneha122 commented on app/server/appsmith-plugins/mssqlPlugin/src/main/java/com/external/plugins/MssqlPlugin.java:
```diff
@@ -363,30 +364,47 @@ public void datasourceDestroy(HikariDataSource connection) {
+                if (auth.getPassword() == null || auth.getPassword().isBlank()) {
                     invalids.add(MssqlErrorMessages.DS_MISSING_PASSWORD_ERROR_MSG);
                 }
+                if (auth.getDatabaseName() == null || auth.getDatabaseName().isBlank()) {
```
Same comment as above

saiprabhu-dandanayak replied:
Resolved

sneha122 commented on app/server/appsmith-plugins/mssqlPlugin/src/main/java/com/external/plugins/MssqlPlugin.java:
```diff
@@ -592,13 +610,27 @@ private static HikariDataSource createConnectionPool(DatasourceConfiguration dat
 
+        try {
+            // Try to establish a connection to validate the configuration
+            DriverManager.getConnection(
```
Why do we have to explicitly do establish connection using DriverManager? When we create HikariDataSource it should automatically send us valid error message with PoolInitializationException, is it not happening now?

akshayvijayjain replied:
@Shirisha-Manthoju

saiprabhu-dandanayak replied:
If we are not explicitly establishing connection using DriverManager , when i give wrong host name like `prabhu` and invalid port like `3434` we are not getting any error.

### when invalid host address is given
![image](https://github.com/user-attachments/assets/0a41b62d-2057-4591-80a3-f9d89bf16e1b)
### when invalid host address and port no are given
![image](https://github.com/user-attachments/assets/14ad3ab3-f968-48fd-9c73-6db008f521b0)

sneha122 replied:
@saiprabhu-dandanayak As we can see in above screenshot, we do get errors when no port and invalid host and port is entered, I am not understanding why establishing connection with driver manager is required. If you check the snowflake connection code [here](https://github.com/appsmithorg/appsmith/blob/ae9742194a558d673e8b5b559ff950448a16e8a6/app/server/appsmith-plugins/snowflakePlugin/src/main/java/com/external/plugins/SnowflakePlugin.java#L171), This also behaves in the similar manner, when we create a new datasource using HikariDataSource constructor. It tries establishing connection with underlying database, if it fails it throws an exception called PoolInitializationException with error message from the driver. We can simply then show this error message in the UI, we wont have to establish connection using driverManager, I hope this makes sense. Please check out the snowflake plugin code as well for reference

saiprabhu-dandanayak replied:
ok , tq for your feedback.

saiprabhu-dandanayak commented on app/server/appsmith-plugins/mssqlPlugin/src/main/java/com/external/plugins/MssqlPlugin.java:
```diff
@@ -363,30 +364,47 @@ public void datasourceDestroy(HikariDataSource connection) {
+        }
+
+        private void validateEndpoints(DatasourceConfiguration datasourceConfiguration, Set<String> invalids) {
+            if (StringUtils.isEmpty(datasourceConfiguration.getUrl())
```
Here getUrl() is a combination of host address and port number , if a user does not provide and clicks on `Test  configuration` an error will be popup.
![image](https://github.com/user-attachments/assets/4630619e-79b2-4f86-af8c-e10f97193344)

sneha122 commented on app/server/appsmith-plugins/mssqlPlugin/pom.xml:
```diff
@@ -16,6 +16,12 @@
 
     <dependencies>
 
+     <dependency>
```
Why is this dependency required?

saiprabhu-dandanayak replied:
i have added commons-lang3 dependency to take `StringUtils.isBlank()` method

sneha122 commented on app/server/appsmith-plugins/mssqlPlugin/src/main/java/com/external/plugins/MssqlPlugin.java:
```diff
@@ -188,121 +202,125 @@ public Mono<ActionExecutionResult> executeCommon(
-                            closeConnectionPostExecution(resultSet, statement, preparedQuery, sqlConnectionFromPool);
-                        }
+                // Log HikariCP status
+                logHikariCPStatus(
```
@saiprabhu-dandanayak Could we please run `mvn spotless:apply` on server code once?

saiprabhu-dandanayak replied:
ok **@sneha122**

sneha122 commented:
Hello @saiprabhu-dandanayak Any updates on the code review comments? I am still seeing changes in pom.xml

saiprabhu-dandanayak commented:
Hi @sneha122  , i have removed explicit `Driver Manager` connection and added updated the code in the `catch` block to raise exceptions and error alerts

sneha122 commented:
> Hi @sneha122 , i have removed explicit `Driver Manager` connection and added updated the code in the `catch` block to raise exceptions and error alerts

Can you please check the pom.xml once? I am still seeing changes in it. Can you please check once what formatter you are using? I am seeing changes to existing code as well where its getting cut off to next line, I believe you may have a line limit formatter on which is causing longer lines to get cut into multiple lines, we don't want to do that.

NilanshBansal commented:
@saiprabhu-dandanayak any updates on this? We haven't heard back from you since the last 2 weeks. 
This PR will be closed in the next 7 days if there are no more updates.

sneha122 commented on app/server/appsmith-plugins/mssqlPlugin/src/main/java/com/external/plugins/MssqlPlugin.java:
```diff
@@ -591,14 +633,45 @@ private static HikariDataSource createConnectionPool(DatasourceConfiguration dat
+                SQLServerException sqlException = (SQLServerException) cause;
+                String sqlState = sqlException.getSQLState();
+
+                if ("08S01".equals(sqlState)) {
```
@saiprabhu-dandanayak Can you please create error constants for such errors? instead of using magic strings in the code

sneha122 commented on app/server/appsmith-plugins/mssqlPlugin/src/main/java/com/external/plugins/MssqlPlugin.java:
```diff
@@ -591,14 +633,45 @@ private static HikariDataSource createConnectionPool(DatasourceConfiguration dat
+
+                    throw new AppsmithPluginException(
+                            AppsmithPluginError.PLUGIN_DATASOURCE_ARGUMENT_ERROR,
+                            "Could not connect to server. Check host and port.",
```
@saiprabhu-dandanayak Can we please check how we have defined error messages in case of postgres plugin and follow the same strategy here? In postgres we are also using something call PSQLState object for error codes, can we follow similar approach here?

sneha122 commented on app/server/appsmith-plugins/mssqlPlugin/src/main/java/com/external/plugins/MssqlPlugin.java:
```diff
@@ -607,7 +680,8 @@ private static HikariDataSource createConnectionPool(DatasourceConfiguration dat
             DatasourceConfiguration datasourceConfiguration, StringBuilder urlBuilder) throws AppsmithPluginException {
         /*
-         * - Ideally, it is never expected to be null because the SSL dropdown is set to a initial value.
+         * - Ideally, it is never expected to be null because the SSL dropdown is set to
```
@saiprabhu-dandanayak Can you please fix the indentation here? It is messing up the code and makes it difficult to understand the actual changes made

sneha122 commented:
Hi @saiprabhu-dandanayak, Any updates on this?

I have created shadow PR here for running test cases: https://github.com/appsmithorg/appsmith/pull/36603

NilanshBansal commented:
Closing this PR due to no response. 
cc: @appsmithorg/query-js-pod
